rfcs/2018-11-23-annotations.md

```diff
@@ -0,0 +1,406 @@
+    ],
+    “metadata”: null
+}
+```
+This annotation set would be sent to core, which would then forward it
+to the frontend to display as desired.
+
+## Questions / considerations
+
+A number of things have come up while drafting this proposal that are
+worth noting.
+
+#### metadata handling
+
+We probably want (and have assumed having,elsewhere in this proposal)
+some way to associate metadata with a _set_ of annotations;
+for instance 'find' annotations might include the total
+number of results and their approximate locations. This does add some
+complexity, however, and it isn't totally clear that it makes sense as
+part of this API; it might make more sense to have a separate 'metadata'
+API, which could provide metadata for certain types of annotations as
+well as other purposes (for instance, this might be a better way of
+communicating when a buffer has unsaved changes.)
+
+#### zero-width annotations
+If we would like to indicate a 'removed range' in a version control
+plugin, it would make sense to communicate this as a zero-width
+annotation. This will also be necessary for selections. So less
+a question than something to be explicit about.
+
+#### type checking and validation
+regardless of how we actually represent annotations in core,
+we would like to have some mechanism for
+validating them (at least those of known types) so that we aren't
+forwarding malformed data blindly to the frontend.
+
+#### fetching ranges of annotations
+Since we only send the visible annotations to the frontend,
+we may want to expose some sort of API to allow fetching annotations
+for other regions of the document.
+This is definitely 'build it when you need it' territory.
+
+#### More efficient invalidation
+We may find that in use we end up with lots of redundant annotations
+in the update RPC; for instance when the user is actively editing
+and/or moving through a document we may find ourselves resending
+large numbers of unchanged annotations. If this ends up being
+the case, we may want to explore more explicit invalidation strategies.
+
+#### Updating annotations in response to edits
+When text is inserted interior of a selection, we expand selection
+to include the new text. When text is inserted in the interior of
+a _style_ span, we split the existing span in two and insert an
+empty span for the inserted text. What should we do when an edit
+occurs inside an annotation?
+
+The 'correct' behaviour is highly dependent on the type of the
+annotation. If we had to choose a single behaviour, I think we would
+want to either destroy modified annotations or to expand them, like
+selections; but we may want to have an 'update strategy' that is
+determined by the annotation's type.
+
+#### Annotations for inter-plugin communication
+A far future concern: but we may want to explore whether annotations
+also make sense as a mechanism for plugins to communicate metadata
+to one another.
+
+#### Sharing annotations between views
```

Comment:
I'm now curious how other editors handle this. Some annotations are clearly per-view (selection, find, autocompletion suggestions). Others I can imagine being per-document (eg breakpoints). I'm curious how other editors handle this; if you have a document open and do "new view", what stuff shows in the new view?